sklearn_numba_dpex/common/kernels.py

```diff
@@ -397,7 +429,149 @@ def partial_sum_reduction(
+    reduction_block_size = 2 * n_sub_group_per_work_group
+
+    # ???: how does this strategy compares to having each thread reducing N contiguous
+    # items ?
+    @dpex.kernel
+    # fmt: off
+    def partial_sum_reduction(
+        summands,    # IN        (n_rows, sum_axis_size)
+        result,      # OUT       (n_rows, math.ceil(size / (2 * reduction_block_size),)
+    ):
+
+        # fmt: on
+        # NB: This kernel only perform a partial reduction
+        group_id = dpex.get_group_id(zero_idx)
+        local_work_id = dpex.get_local_id(zero_idx)
+
+        sum_axis_size = summands.shape[zero_idx]
+        n_block_for_col = result.shape[zero_idx]
+
+        local_block_id_in_col = (group_id % n_block_for_col)
+        first_row_idx = local_block_id_in_col * reduction_block_size
+        local_col_idx = (local_work_id % sub_group_size)
+        local_row_idx = (local_work_id // sub_group_size)
+        col_idx = (
+            group_id // n_block_for_col) * sub_group_size + local_col_idx
+        augend_row_idx = first_row_idx + local_row_idx
+        addend_row_idx = augend_row_idx + n_sub_group_per_work_group
+
+        local_values = dpex.local.array(local_values_size, dtype=dtype)
+
+        # Each work item reads two value in global memory and sum it into the local
+        # memory
+        if (col_idx >= n_cols) or (augend_row_idx >= sum_axis_size):
+            local_values[local_row_idx, local_col_idx] = zero
+        elif addend_row_idx >= sum_axis_size:
+            local_values[local_row_idx, local_col_idx] = fused_unary_func_(
+                summands[augend_row_idx, col_idx])
+        else:
+            local_values[local_row_idx, local_col_idx] = fused_unary_func_(
+                summands[augend_row_idx, col_idx]) + fused_unary_func_(
+                    summands[addend_row_idx, col_idx])
+
+        dpex.barrier(dpex.CLK_LOCAL_MEM_FENCE)
+
+        current_n_nows = n_sub_group_per_work_group
+        for i in range(local_n_iterations):
+            # We discard half of the remaining active subgroups at each iteration
+            current_n_nows = current_n_nows // two_as_a_long
+            if (local_row_idx < current_n_nows) and (
+                    col_idx < n_cols) and (
+                        first_row_idx + local_row_idx + current_n_nows) < sum_axis_size:
+                local_values[local_row_idx, local_col_idx] += local_values[
+                    local_row_idx + current_n_nows, local_col_idx
+                ]
+
+            dpex.barrier(dpex.CLK_LOCAL_MEM_FENCE)
+
+        # At this point local_values[0] + local_values[1] is equal to the sum of all
+        # elements in summands that have been covered by the work group, we write it
+        # into global memory
+        if (local_row_idx == zero_idx) and (col_idx < n_cols):
+            result[local_block_id_in_col, col_idx] = (
+                local_values[zero_idx, local_col_idx] + local_values[
+                    one_idx, local_col_idx])
```

Comment:
again the formatting does not look blackish to me. I would have expected something more readable like:

```suggestion
            result[local_block_id_in_col, col_idx] = (
                local_values[zero_idx, local_col_idx]
                + local_values[one_idx, local_col_idx]
            )
```